### PR TITLE
MEN-4093 Remote terminal overview and configuration options.

### PR DESCRIPTION
**Pull request by merlin-northern**
ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+## mender-shell and the user model
+
+Both the operating system on the devices and Mender support multiple users.
+Any number of these at any moment may start terminals from the UI.
+Mender-shell supports multiple shells of multiple Mender users and requires
+you to specify which system user to run the shell process as. 
+
+An arbitrary number of users each running an arbitrary number of sessions
+can easily lead to resources exhaustion on a device. To this end, mender-shell
+imposes two configurable restrictions: on the total number of shells, and the number
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:64]: I would break these restrictions into a bulleted list.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+of sessions per user.
+
+On the other hand, you can imagine a scenario when you reach
+both limits and lock yourself out from the remote terminal.
+In order to counter act that, mender-shell supports automatic termination
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:69]: In order to protect against this scenario, mender-shell support....

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+of idle and/or expired sessions.
+
+## Sessions and terminals
+
+Despite the fact that mender-shell sessions are not login sessions
+that you may expect from the other solutions, e.g.: SSH, the terminal
+you get is fully functional. Indeed, it is virtually impossible to distinguish
+it from the regular SSH interactive session just by working with the shell.
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:77]: I think this paragraph can be removed.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+Remote terminal session consists of a spawned shell (as a subprocess of
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:79]: A remote terminal session....

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+mender-shell) a unique string called a session id, and a system username
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:80]: subprocess of the mender-shell process)...

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+that is running the shell command. The main purpose of mender-shell
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:81]: Remove "that is running the shell command"
The main responsibility of the mender-shell executable...

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+is to read messages from the websocket, route them by session id
+to the correct shell, read the output going to the terminal,
+and pass it upstream over the websocket.
+
+## Further reading
+
+* For a detailed configuration options please refer to
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:88]: a detailed list of configuration options

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+mender-shell) a unique string called a session id, and a system username
+that is running the shell command. The main purpose of mender-shell
+is to read messages from the websocket, route them by session id
+to the correct shell, read the output going to the terminal,
+and pass it upstream over the websocket.
+
+## Further reading
+
+* For a detailed configuration options please refer to 
+[mender-shell configuration section](../../03.Client-installation/06.Configuration-file/02.Mender-shell/docs.md).
```
Comment (drewmoseley) [inline 02.Overview/14.Remote.Terminal/docs.md:89]: the mender-shell configuration section.

File: 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+```
+
+####	User
+
+The name of a user that mender-shell will run the shell as. Must resolve to valid
+user id and must posses a valid group id. This a mandatory field.
+
+####	Terminal
+
+Allows you to provide terminal settings.
```
Comment (drewmoseley) [inline 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md:68]: Example needed here

File: 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+```
+    "Terminal": {
+        "Height": 24,
+        "Width": 80
+    }
+```
+
+####	Sessions
+
+User sessions settings.
```
Comment (drewmoseley) [inline 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md:89]: Example needed here

File: 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+##### StopExpired
+
+It controls whether to stop sessions. If set to `false` mender-shell will
```
Comment (drewmoseley) [inline 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md:93]: Determines whether stopped sessions will be automatically expired and removed.

**Review by drewmoseley — CHANGES_REQUESTED** [review]

File: 02.Overview/01.Introduction/docs.md [hunk 12]
```diff
@@ -7,6 +7,8 @@ taxonomy:
 Mender is a secure and robust software update system designed to handle
 large number of devices. Mender is designed according to a simple client/server
 architecture, allowing central management of deployments to all devices.
+It also supports an add-on mechanism, which you can use to open an interactive
+remote terminal session to any device under Mender control.
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/01.Introduction/docs.md:11]: currently this is the only and the first add-on. And it is not very generic at the moment. I want to mention it here, since it is an important architectural change and a new significant feature. This was my attempt, if you have any suggestions into which direction this should be changed, I am all ears.

**Comment by mirzak** [comment]
Have not yet looked close enough on content but more interested on architecture of the docs, and here @estenberg should have a say.

I am not super convinced that we should add this to `Overview` and `Client installation` sections. Instead I am wondering if we should have a separate "Add-on" top level category or something similar to keep it isolated from the rest and over time we might merge this more, but our current docs focused on OTA, and just adding non-related blocks to existing section I find confusing.

BTW, I have the site rendered here, http://13.53.78.221:8080/development/overview

File: 02.Overview/01.Introduction/docs.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/01.Introduction/docs.md:11]: fixed the best I could.

File: 02.Overview/01.Introduction/docs.md [hunk 13]
```diff
@@ -111,3 +113,19 @@ Example using the DFU protocol to update an external microcontroller:
 
 You can find more information on above example in the
 [How to do a proxy deployment to an FRDM-K64F](https://hub.mender.io/t/how-to-do-a-proxy-deployment-to-an-frdm-k64f-device-connected-to-a-raspberry-pi-3/1619?target=_blank) tutorial.
+
+
+## The architecture of a Mender add-on
+
+Mender supports add-ons in a form of standalone binaries exchanging messages
+with the client over [DBus](https://dbus.freedesktop.org/doc/api/html/).
+One example being [mender-shell](https://github.com/mendersoftware/mender-shell)
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/01.Introduction/docs.md:122]: fixed.

File: 02.Overview/01.Introduction/docs.md [hunk 14]
```diff
@@ -111,3 +113,19 @@ Example using the DFU protocol to update an external microcontroller:
...
+which is an implementation of remote terminal access to devices.
+The below picture shows the architecture of the solution as you can
+see it from the device perspective.
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/01.Introduction/docs.md:125]: sure.

File: 02.Overview/01.Introduction/docs.md [hunk 15]
```diff
@@ -111,3 +113,19 @@ Example using the DFU protocol to update an external microcontroller:
...
+
+![addon-architecture](addon-architecture-device.png)
+
+Please refer to
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/01.Introduction/docs.md:129]: fixed.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+With Mender, you can start an interactive shell to any accepted
+device by clicking "Launch a new Terminal" in the UI (see the picture
+below).
+
+![launch-termninal](launch-new-termninal.png)
+
+All you need is 
+[mender-shell](https://github.com/mendersoftware/mender-shell)
+configured and running alongside the Mender Client, and you can
+get a live terminal where you can freely type commands, as the below
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:18]: dropped.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+Remote Terminal by Mender passes messages of a defined protocol over
+a websocket connection. The two main actors: the mender-shell and Mender UI
+communicate using a well-defined open protocol over a websocket connection.
+They exchange messages which carry user keystrokes, terminal output,
+and control data. Therefore, there is no relation to SSH and no part
+of it is either used or required. You are, however, able to run sshd or ssh
+command, since you have a fully functional terminal at your disposal.
+Mender does not impose any additional restrictions on the commands,
+shell, or terminal. Below picture shows a simplified architecture
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:36]: fixed.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+## mender-shell and the rest of the Mender ecosystem
+
+The coupling between mender-shell and the rest of Mender is not tight.
+Remote terminal expects a certain well-defined DBus API to be in place
+on a device, and a working websocket connection endpoint; none of these
+have to come from or go to Mender.
+As long as you send compatible messages over the websocket, mender-shell will work.
+With Mender, however, you have everything in place, and all you need
+is to perform a simple configuration of the client and start mender-shell,
+to enjoy interactive sessions from the Mender UI to your devices. Additionally,
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:51]: ok.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+in this picture you can take advantage of the Role Based Access Control
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:52]: right.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+and Audit Logs, since we fully integrated mender-shell add-on into Mender.
+
+## mender-shell and the user model
+
+Both the operating system on the devices and Mender support multiple users.
+Any number of these at any moment may start terminals from the UI.
+Mender-shell supports multiple shells of multiple Mender users and requires
+you to specify which system user to run the shell process as.
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:60]: rephrased to escape the passive:
```
 The mender-shell supports simultaneous shells from different Mender users.
 On the device, you run the shell as a system user which you can set
 in the configuration.
```

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 1 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:64]: right. done.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 2 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:69]: fixed.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 3 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:77]: removed.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 4 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:79]: right. (I meant the Remote terminal as a proper name of the feature). fixed.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 5 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:80]: fixed.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 6 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:81]: fixed.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 7 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:88]: fixed.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 8 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:89]: fixed

File: 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md (hunk 9 above)
Comment (merlin-northern) [reply 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md:68]: added.

File: 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md (hunk 10 above)
Comment (merlin-northern) [reply 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md:89]: it is there: below :)

File: 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md (hunk 11 above)
Comment (merlin-northern) [reply 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md:93]: fixed.

**Comment by merlin-northern** [comment]
> Have not yet looked close enough on content but more interested on architecture of the docs, and here @estenberg should have a say.
> 
> I am not super convinced that we should add this to `Overview` and `Client installation` sections. Instead I am wondering if we should have a separate "Add-on" top level category or something similar to keep it isolated from the rest and over time we might merge this more, but our current docs focused on OTA, and just adding non-related blocks to existing section I find confusing.

I did that on purpose: to hint that we are changing.
 
> BTW, I have the site rendered here, http://13.53.78.221:8080/development/overview

thanks.

**Comment by merlin-northern** [comment]
@estenberg @mirzak @drewmoseley see also: https://github.com/mendersoftware/mender-docs/pull/1241 (related, and the discussion there)

File: 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md (hunk 10 above)
Comment (drewmoseley) [reply 03.Client-installation/06.Mender-shell-configuration-file/50.Mender-shell-configuration-options/docs.md:89]: Oh, I didn't notice the header levels here indicating that the below were sub-fields

**Review by drewmoseley — APPROVED** [review]
LGTM

**Comment by mirzak** [comment]
Can this be rebased on top of https://github.com/mendersoftware/mender-docs/tree/feature-deviceconnect, so that we get everything related in one place?

**Comment by kjaskiewiczz** [comment]
shouldn't a target of this PR be the feature-deviceconnect branch?

**Comment by merlin-northern** [comment]
> shouldn't a target of this PR be the feature-deviceconnect branch?

sure, but I think the branch was no existent at the time.

**Comment by merlin-northern** [comment]
> Can this be rebased on top of https://github.com/mendersoftware/mender-docs/tree/feature-deviceconnect, so that we get everything related in one place?

sure.

**Comment by merlin-northern** [comment]
> Can this be rebased on top of https://github.com/mendersoftware/mender-docs/tree/feature-deviceconnect, so that we get everything related in one place?

it was supposed to be the base, thats why I had taken it early.

**Comment by estenberg** [comment]
Thanks @merlin-northern !
I think this content is great! I do have a few high-level comments (sorry!):

* Creating a new top-level Chapter called "Add on" is a good idea I think. Basically you just need to move the Overview/Remote terminal and the Client installation/Mender shell as subsections of it. The main reason is that this feature will be part of "Troubleshoot" so there will be more of these features going forward (file up/download, etc.). It should be clear they can be disabled/removed, and there will be a commercial add-on offering around this too - and it would be nice if the marketing were aligned with the docs structure.

* Taxonomy is a bit mixed, I checked with Farshad and "Remote terminal" should be what we use. Now it's a mix of Remote terminal, remote access, shell.

I will provide some comments in the PR as well.

File: 02.Overview/01.Introduction/docs.md [hunk 21]
```diff
@@ -7,6 +7,9 @@ taxonomy:
 Mender is a secure and robust software update system designed to handle
 large number of devices. Mender is designed according to a simple client/server
 architecture, allowing central management of deployments to all devices.
+It also supports an add-on mechanism, which currently has one implementation
```
Comment (estenberg) [inline 02.Overview/01.Introduction/docs.md:10]: Would rephrase slightly to, "It also supports an add-on features, enabling support for related device management features like Remote terminal."

File: 02.Overview/01.Introduction/docs.md [hunk 22]
```diff
@@ -111,3 +114,18 @@ Example using the DFU protocol to update an external microcontroller:
 
 You can find more information on above example in the
 [How to do a proxy deployment to an FRDM-K64F](https://hub.mender.io/t/how-to-do-a-proxy-deployment-to-an-frdm-k64f-device-connected-to-a-raspberry-pi-3/1619?target=_blank) tutorial.
+
+
+## The architecture of a Mender add-on
```
Comment (estenberg) [inline 02.Overview/01.Introduction/docs.md:119]: This is good, would leave it here.

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 23]
```diff
@@ -0,0 +1,86 @@
+---
+title: Remote Terminal
+taxonomy:
+category: docs
+---
+
+## Remote access
```
Comment (estenberg) [inline 02.Overview/14.Remote.Terminal/docs.md:7]: Remove this heading - doesn't add much?

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+
+With Mender, you can start an interactive shell to any accepted
```
Comment (estenberg) [inline 02.Overview/14.Remote.Terminal/docs.md:9]: interactive shell -> terminal session?
accepted -> connected?

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+device by clicking "Launch a new Terminal" in the UI (see the picture
+below).
+
+![launch-termninal](launch-new-termninal.png)
+
+All you need is 
+[mender-shell](https://github.com/mendersoftware/mender-shell)
```
Comment (estenberg) [inline 02.Overview/14.Remote.Terminal/docs.md:16]: This (link to github) is a bit raw? Perhaps link to the other section about it in the docs instead?

Comment (estenberg) [reply 02.Overview/14.Remote.Terminal/docs.md:16]: Maybe there is no better place to link currently? I would expect some kind of setup instructions here though?

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+configured and running alongside the Mender Client, and you can
+get a live terminal where you can freely type commands.
+
+![whoami-mender](whoami-mender.png)
+
+Every keystroke in the above popup window will go to the standard input
+of a shell running on the device.
+
+## mender-shell vs SSH
```
Comment (estenberg) [inline 02.Overview/14.Remote.Terminal/docs.md:25]: "No relation to SSH"?

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+Mender does not impose any additional restrictions on the commands,
+shell, or terminal. The picture below shows a simplified architecture
+with a websocket, mender-shell, and a shell running in a child process
+of mender-shell with allocated pseudo tty.
+
+![mender-shell-and-ws](mender-shell-and-ws.png)
+
+## mender-shell and the rest of the Mender ecosystem
+
+The coupling between mender-shell and the rest of Mender is not tight.
```
Comment (estenberg) [inline 02.Overview/14.Remote.Terminal/docs.md:43]: Perhaps rephrase "There is loose coupling between Mender and add-ons like mender-shell."

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+of it is either used or required. You are, however, able to run sshd or ssh
+command, since you have a fully functional terminal at your disposal.
+Mender does not impose any additional restrictions on the commands,
+shell, or terminal. The picture below shows a simplified architecture
+with a websocket, mender-shell, and a shell running in a child process
+of mender-shell with allocated pseudo tty.
+
+![mender-shell-and-ws](mender-shell-and-ws.png)
+
+## mender-shell and the rest of the Mender ecosystem
```
Comment (estenberg) [inline 02.Overview/14.Remote.Terminal/docs.md:41]: "Loose coupling to Mender" ?

File: 02.Overview/14.Remote.Terminal/docs.md [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+Remote terminal expects a certain well-defined DBus API to be in place
+on a device, and a working websocket connection endpoint; none of these
+have to come from or go to Mender.
+As long as you send compatible messages over the websocket, mender-shell will work.
+With Mender, however, you have everything in place, and all you need
+is to perform a simple configuration of the client and start mender-shell.
+Additionally, you can take advantage of the Role Based Access Control
+and Audit Logs, since we fully integrated mender-shell add-on into Mender.
+
+## mender-shell and the user model
```
Comment (estenberg) [inline 02.Overview/14.Remote.Terminal/docs.md:53]: "Support for multiple users"?

**Review by estenberg — CHANGES_REQUESTED** [review]

File: 02.Overview/01.Introduction/docs.md (hunk 21 above)
Comment (merlin-northern) [reply 02.Overview/01.Introduction/docs.md:10]: right.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 23 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:7]: sure.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 24 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:9]: sure. both sound better. I put "accepted" because technically it is most correct: you cant run a remote terminal to a device that you have not accepted.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 25 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:16]: got me with both comments. at the time of writing there was nothing. so I tried to link to anything. fixed to link to donwloads.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 26 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:25]: sure

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 27 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:43]: I exactly had in mind "loose" but somehow I decided otherwise. fixed.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 28 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:41]: sure.

File: 02.Overview/14.Remote.Terminal/docs.md (hunk 29 above)
Comment (merlin-northern) [reply 02.Overview/14.Remote.Terminal/docs.md:53]: sure.

File: 03.Add-ons/chapter.md [hunk 30]
```diff
@@ -0,0 +1,15 @@
...
+title: Add-ons
+taxonomy:
+    category: docs
+---
+
+### Chapter 3
+
+# Add-ons
+
+Mender supports extensions in a form of standalone binaries running
```
Comment (estenberg) [inline 03.Add-ons/chapter.md:11]: "extensions" -> "add-ons"

File: 03.Add-ons/chapter.md [hunk 31]
```diff
@@ -0,0 +1,15 @@
...
+on a device along side the Mender client. We call them _add-ons_. They
```
Comment (estenberg) [inline 03.Add-ons/chapter.md:12]: Can remove "We call them _add-ons_. "

File: 03.Add-ons/chapter.md (hunk 30 above)
Comment (estenberg) [reply 03.Add-ons/chapter.md:11]: Can emphasise "add-ons"

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+
+Delivering commands securely, maintaining the identity of the communication
+endpoints, ensuring endpoint authentication, are critical factors in a secure
+remote command execution. This section gives a brief security overview
+of the Remote Terminal add-on.
+
+## Remote Terminal
+
+With Mender, you can start an interactive shell to any accepted device by clicking "Launch a new Terminal" in the UI.
+The name of the feature is [Remote Terminal](../10.Overview/docs.md).
```
Comment (estenberg) [inline 03.Add-ons/01.Remote-Terminal/40.Security/docs.md:15]: This is implicit, can remove this sentence.

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+Remote Terminal is an optional feature.
+One part of the Remote terminal configuration is the username of the user on the device for whom the shell will be created.
+Using Remote Terminal you can execute any command the user on the device can execute.
+Like with any other operation, the Mender backend will allow only authenticated Mender users to access the Remote Terminal.
+You can apply additional restrictions on which Mender users can access Remote Terminal using RBAC.
+The [mender-shell](https://github.com/mendersoftware/mender-shell), part of the Remote Terminal that is running on the device,
+obtains device token through DBus API. It is the same token Mender client is using.
+<!--AUTOVERSION: "https://tools.ietf.org/html/rfc6455#section-%"/ignore-->
+The mender-shell uses [Encrypted WebSocket connections](https://tools.ietf.org/html/rfc6455#section-11.1.2) to communicate with the server.
```
Comment (estenberg) [inline 03.Add-ons/01.Remote-Terminal/40.Security/docs.md:24]: mender-shell -> "Remote terminal"

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md [hunk 34]
```diff
@@ -0,0 +1,29 @@
...
+In general, the security impact of enabling Remote Terminal is similar to the one when enabling SSH.
+The connection to the device is secure, but there are no limitations on what command the user can execute,
+except for the user permissions on the device.
+There are also some important differences between enabling Remote Terminal and enabling SSH server on the device.
+Remote Terminal does not opens or listens on any port and the device initiates the connection.
```
Comment (estenberg) [inline 03.Add-ons/01.Remote-Terminal/40.Security/docs.md:29]: "opens" -> "open"

Comment (estenberg) [reply 03.Add-ons/01.Remote-Terminal/40.Security/docs.md:29]: "listens" -> "listen"

Comment (estenberg) [reply 03.Add-ons/01.Remote-Terminal/40.Security/docs.md:29]: Recommending "Remote Terminal does not open nor listen on any port: the device initiates the connection to the server."

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+taxonomy:
+    category: docs
+---
+
+Delivering commands securely, maintaining the identity of the communication
+endpoints, ensuring endpoint authentication, are critical factors in a secure
+remote command execution. This section gives a brief security overview
+of the Remote Terminal add-on.
+
+## Remote Terminal
```
Comment (estenberg) [inline 03.Add-ons/01.Remote-Terminal/40.Security/docs.md:12]: Would mention audit logs at the end.

Also would consider having subsections here to improve structure.

E.g.

"Local device user"

"Role Based Access Control"

"No open ports"

"Audit logs"

Finally, need to tag that some of these features are only available if they are using Mender Enterprise (RBAC, Audit logs).

**Review by estenberg — APPROVED** [review]
Overall very nice, I have some nitpicks to improve the security section in particular but it can be improved later too.

**Comment by merlin-northern** [comment]
> Overall very nice, I have some nitpicks to improve the security section in particular but it can be improved later too.

great! it is not only not mine, but belongs to KJ, but also already reviewed ;-) I will correct according to your comments of course.

File: 03.Add-ons/chapter.md (hunk 30 above)
Comment (merlin-northern) [reply 03.Add-ons/chapter.md]: right.

File: 03.Add-ons/chapter.md (hunk 31 above)
Comment (merlin-northern) [reply 03.Add-ons/chapter.md]: right.

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md (hunk 32 above)
Comment (merlin-northern) [reply 03.Add-ons/01.Remote-Terminal/40.Security/docs.md]: removed.

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md (hunk 33 above)
Comment (merlin-northern) [reply 03.Add-ons/01.Remote-Terminal/40.Security/docs.md]: fixed.

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md (hunk 34 above)
Comment (merlin-northern) [reply 03.Add-ons/01.Remote-Terminal/40.Security/docs.md]: recommendation taken.

File: 03.Add-ons/01.Remote-Terminal/40.Security/docs.md (hunk 35 above)
Comment (merlin-northern) [reply 03.Add-ons/01.Remote-Terminal/40.Security/docs.md:12]: added subsections, noted about enterprise features, and corrected.

**Review by estenberg — APPROVED** [review]

**Comment by merlin-northern** [comment]
thanks so much @estenberg
